### PR TITLE
feat(serve): support --consolelogs option

### PR DESCRIPTION
**Pull request by tlancina**
Sets up a websocket server and injects a script in to the app to
pass console logs back to the server to be printed in the terminal.

--consolelogs-port configures the websocket server port

### CODE REVIEW

File: assets/log-server.ts [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+
+export function isConsoleLogServerMessage(m: any): m is ConsoleLogServerMessage {
+  return m
+    && typeof m.category === 'string'
+    && typeof m.type === 'string'
+    && m.data && typeof m.data.length === 'number';
+}
+
+export async function createConsoleLogServer(host: string, port: number): Promise<ζws.Server> {
+  const WebSocket = await import('ws');
```
Comment (dwieeb) [inline assets/log-server.ts]: Can you add this to a top-level import?

File: assets/log-server.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+      }
+
+      if (!isConsoleLogServerMessage(msg)) {
+        const m = util.inspect(msg, { colors: chalk.enabled });
+        process.stderr.write(`Bad format in client message: ${m}\n`);
+        return;
+      }
+
+      if (msg.category === 'console') {
+        let status: Chalk | undefined; // unknown levels are normal color
```
Comment (dwieeb) [inline assets/log-server.ts]: Sorry to be picky--can we use the devkit colors? https://github.com/angular/angular-cli/blob/master/packages/angular_devkit/core/src/terminal/colors.ts

Should look like this:

```ts
import { terminal } from '@angular-devkit/core';

status = terminal.red
```

File: builders/cordova-build/schema.d.ts [hunk 3]
```diff
@@ -5,4 +5,6 @@ export interface CordovaBuildBuilderSchema {
   sourceMap?: boolean;
   cordovaAssets?: boolean;
   cordovaMock?: boolean;
+  consolelogs?: boolean;
+  consolelogsPort?: number;
```
Comment (dwieeb) [inline builders/cordova-build/schema.d.ts]: 👌

File: builders/cordova-build/schema.json [hunk 4]
```diff
@@ -7,6 +7,15 @@
       "type": "string",
       "description": "Target to build."
     },
+    "consolelogs": {
+      "type": "boolean",
+      "description": "Inject script to print app console logs to Ionic CLI"
```
Comment (dwieeb) [inline builders/cordova-build/schema.json]: Can we just say "Inject script to print console logs to the terminal"?

File: builders/cordova-serve/index.ts [hunk 5]
```diff
@@ -52,6 +53,14 @@ class CordovaDevServerBuilder extends DevServerBuilder {
     super(context);
   }
 
+  run(builderConfig: BuilderConfiguration<DevServerBuilderOptions>): Observable<BuildEvent> {
+    if (this.cordovaBuildOptions.consolelogs && this.cordovaBuildOptions.consolelogsPort) {
+      createConsoleLogServer(builderConfig.options.host, this.cordovaBuildOptions.consolelogsPort)
```
Comment (dwieeb) [inline builders/cordova-serve/index.ts]: If you wanted to get a little fancier, you can use `fromPromise`: http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#static-method-fromPromise

File: assets/log-server.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
+import chalk, { Chalk } from 'chalk';
```
Comment (dwieeb) [inline assets/log-server.ts]: I'm not sure I think of this file as an "asset"... maybe move back into `cordova-serve`? Then your `.gitignore` cleans up nicely too

File: builders/cordova-build/schema.json (hunk 4 above)
Comment (tlancina) [reply builders/cordova-build/schema.json]: Yes I thought this sounded weird/incorrect, but it's what the CLI currently says: https://github.com/ionic-team/ionic-cli/blob/6edf37367473cb3726cfefcd570e0a04b7fae461/packages/ionic/src/lib/project/ionic1/serve.ts#L24
https://github.com/ionic-team/ionic-cli/blob/6edf37367473cb3726cfefcd570e0a04b7fae461/packages/ionic/src/lib/project/ionic-angular/serve.ts#L32

I'll change it in the CLI PR as well: https://github.com/ionic-team/ionic-cli/pull/3900/files#diff-43bee07de60dee9e18ddd7a7fbc66673R42

File: builders/cordova-serve/index.ts (hunk 5 above)
Comment (tlancina) [reply builders/cordova-serve/index.ts]: I didn't want to actually slow the build down waiting for the server to spin up by chaining it with the call to `super`, is that what you meant? Something like `from(createConsoleLogServer(...)).pipe(_ => super.run(builderConfig))`? My observable foo isn't great. 

I'm going to leave it as is for now so let me know if you had something different in mind.

File: package.json [hunk 7]
```diff
@@ -31,8 +31,10 @@
   ],
   "dependencies": {
     "@schematics/angular": "^7.0.3",
+    "chalk": "^2.4.2",
```
Comment (dwieeb) [inline package.json]: Shouldn't need this anymore, right?

Comment (tlancina) [reply package.json]: Good catch!

**Pull request merged**